### PR TITLE
Hb user wed

### PR DESCRIPTION
**Pull request by fyarkiy**

### CODE REVIEW

File: src/main/java/com/cinema/exception/PasswordEncryptionException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.cinema.exception;
+
+public class PasswordEncryptionException extends RuntimeException {
+    public PasswordEncryptionException(String message, Throwable ex) {
```
Comment (Nokafan) [inline src/main/java/com/cinema/exception/PasswordEncryptionException.java:4]: Throwable - is it correct?

File: src/main/java/com/cinema/security/AuthenticationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @Override
+    public User register(String email, String password) {
+        byte[] salt = HashUtil.getSalt();
+        User user = new User();
+        user.setPassword(HashUtil.hashPassword(password,salt));
+        user.setEmail(email);
+        user.setSalt(salt);
+        return userService.add(user);
```
Comment (Nokafan) [inline src/main/java/com/cinema/security/AuthenticationServiceImpl.java:27]: We have to salt and hash password in UserServiceImpl

File: src/main/java/com/cinema/exception/PasswordEncryptionException.java (hunk 1 above)
Comment (dariiabutorina) [reply src/main/java/com/cinema/exception/PasswordEncryptionException.java:4]: Let's change this

File: src/main/java/com/cinema/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.Optional;
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+
+    @Override
+    public User add(User user) {
+        return userDao.add(user);
```
Comment (dariiabutorina) [inline src/main/java/com/cinema/service/impl/UserServiceImpl.java:20]: All the business logic somehow related to initializing or setting the values to our fields must be on the service layer.
If somebody creates the object _User_ not throughout the _register_ method but using the _userService.create()_ instead, we'll not have its password hashed, which might cause some mistakes in the future while working with the object or with the security itself.
So let's move the hashing logic to this method.

**Review by dariiabutorina — CHANGES_REQUESTED** [review]

File: src/main/java/com/cinema/exception/PasswordEncryptionException.java (hunk 1 above)
Comment (fyarkiy) [reply src/main/java/com/cinema/exception/PasswordEncryptionException.java:4]: Any ideas - change to what?

Comment (dariiabutorina) [reply src/main/java/com/cinema/exception/PasswordEncryptionException.java:4]: Sorry, it's okay :D
Got confused

**Review by dariiabutorina — APPROVED** [review]

File: src/main/java/com/cinema/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+
+    @Override
+    public User add(User user) {
+        byte[] salt = HashUtil.getSalt();
+        user.setSalt(salt);
+        user.setPassword(HashUtil.hashPassword(user.getPassword(),salt));
```
Comment (RomanDubovskyi) [inline src/main/java/com/cinema/service/impl/UserServiceImpl.java]:
```suggestion
        user.setSalt(HashUtil.getSalt());
        user.setPassword(HashUtil.hashPassword(user.getPassword(), user.getSalt());
```

File: src/main/java/com/cinema/dao/impl/UserDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> userQuery =
+                    session.createQuery("from User u where u.email =: email", User.class);
+            userQuery.setParameter("email", email);
+            return userQuery.uniqueResultOptional();
+        } catch (Exception e) {
+            throw new DataProcessingException("can't find user with email " + email, e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/cinema/dao/impl/UserDaoImpl.java]:
```suggestion
            throw new DataProcessingException("Exception while getting user with email " + email, e);
```

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**